### PR TITLE
Add tooling to CodeStyle layer to generate global configs

### PR DESCRIPTION
**Pull request by mavasani**
Implements the post-build tooling support for https://github.com/dotnet/roslyn/issues/49250#issuecomment-736879786. The tool generates and packages the following files into the CodeStyle package:

1) Global analyzer config files to enable/disable bucket of code style rules based on user specified AnalysisLevel/AnalysisMode
2) Targets file to map the user specified AnalysisLevel/AnalysisMode to the appropriate global analyzer config and pass it down to the compiler

Note that the post-build tool is primarily extracted from https://github.com/dotnet/roslyn-analyzers/blob/master/src/Tools/GenerateDocumentationAndConfigFiles/Program.cs.

~~**TODO**: Add a new enum parameter `EnforceOnBuild` to CodeStyle analyzer constructor, which will determine the correct bucketing for code style analyzers for build enforcement. This will be done in a follow-up PR~~ https://github.com/dotnet/roslyn/pull/49995

### CODE REVIEW

**Comment by mavasani** [comment]
Also tagging @Evangelink and @Youssef1313 for reviews - they have been helping implement and review the corresponding post-build tooling in roslyn-analyzers repo.

File: src/CodeStyle/Tools/Program.cs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            if (args.Length != ExpectedArguments)
+            {
+                Console.Error.WriteLine($"Excepted {ExpectedArguments} arguments, found {args.Length}: {string.Join(';', args)}");
+                return 1;
+            }
+
+            var language = args[0];
+            var outputDir = args[1];
+            var targetsFileName = args[2];
+            var assemblyList = args[3].Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries).ToList();
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:38]: Why the `ToList`? Working with the array should be enough here.

File: src/CodeStyle/Tools/Program.cs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+            CreateGlobalConfigFiles(language, outputDir, assemblyList);
+            CreateTargetsFile(language, outputDir, targetsFileName);
+            return 0;
+        }
+
+        private static void CreateGlobalConfigFiles(string language, string outputDir, List<string> assemblyList)
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:45]: You could replace `List` with `IEnumerable`.

File: src/CodeStyle/Tools/Program.cs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                SortedList<string, DiagnosticDescriptor> sortedRulesById)
+            {
+                var text = GetGlobalconfigText(analysisMode, sortedRulesById);
+                var directory = Directory.CreateDirectory(folder);
+                var configFilePath = Path.Combine(directory.FullName, configFileName);
+                File.WriteAllText(configFilePath, text);
+                return;
+
+                // Local functions
+                static string GetGlobalconfigText(AnalysisMode analysisMode, SortedList<string, DiagnosticDescriptor> sortedRulesById)
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:94]: I don't know what's the preference on roslyn repo but local functions of local functions is a little hard on cognitive load.

File: src/CodeStyle/Tools/Program.cs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                    AddRules(analysisMode);
+                    return result.ToString();
+
+                    void StartGlobalconfig()
+                    {
+                        result.AppendLine(@"# NOTE: Requires **VS2019 16.7** or later");
+                        result.AppendLine();
+                        result.AppendLine(@$"# Style rules with '{analysisMode}' analysis mode");
+                        result.AppendLine();
+                        result.AppendLine($@"is_global = true");
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:107]: Un-needed `$@` or there is something missing.

File: src/CodeStyle/Tools/Program.cs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                static string GetGlobalconfigText(AnalysisMode analysisMode, SortedList<string, DiagnosticDescriptor> sortedRulesById)
+                {
+                    var result = new StringBuilder();
+                    StartGlobalconfig();
+                    AddRules(analysisMode);
+                    return result.ToString();
+
+                    void StartGlobalconfig()
+                    {
+                        result.AppendLine(@"# NOTE: Requires **VS2019 16.7** or later");
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:103]: Un-needed `@`.

File: src/CodeStyle/Tools/Program.cs [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+                        result.AppendLine();
+                        result.AppendLine(@$"# Style rules with '{analysisMode}' analysis mode");
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:105]: Un-needed `@`.

File: src/CodeStyle/Tools/Program.cs [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+                        result.AppendLine();
+                        result.AppendLine($@"is_global = true");
+                        result.AppendLine();
+
+                        // Append 'global_level' to ensure conflicts are properly resolved between different global configs:
+                        //   1. Lowest precedence (-1): SDK config file generated by us.
+                        //   2. Highest predence (non-negative integer): User provided config.
+                        // See https://github.com/dotnet/roslyn/issues/48634 for further details.
+                        result.AppendLine($@"global_level = -1");
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:114]: Un-needed `$@`.

File: src/CodeStyle/Tools/Program.cs [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                                // Rule had the same default severity and enabled state in the release.
+                                // We do not need to generate any entry.
+                                return false;
+                            }
+
+                            var severityString = GetRuleSeverity(isEnabledByDefault, severity);
+
+                            result.AppendLine();
+                            result.AppendLine($"# {rule.Id}: {rule.Title}");
+                            result.AppendLine($@"dotnet_diagnostic.{rule.Id}.severity = {severityString}");
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:148]: Un-needed `@`.

File: src/CodeStyle/Tools/Program.cs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                                    isEnabledInNonDefaultMode = rule.CustomTags.Contains(s_highlyRecommendedTag) || rule.CustomTags.Contains(s_recommendedTag);
+                                    break;
+
+                                case AnalysisMode.Default:
+                                    // Retain the default severity and enabled by default values.
+                                    isEnabledInNonDefaultMode = false;
+                                    break;
+
+                                default:
+                                    throw new NotSupportedException();
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:182]: Adding some message would help to understand what's going-on if the issue does happen.

File: src/CodeStyle/Tools/Program.cs [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                        static string GetRuleSeverity(bool isEnabledByDefault, DiagnosticSeverity defaultSeverity)
+                        {
+                            if (isEnabledByDefault)
+                            {
+                                return GetSeverityString(defaultSeverity);
+                            }
+                            else
+                            {
+                                return GetSeverityString(null);
+                            }
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:198]:
```suggestion
                            return isEnabledByDefault
                            	? GetSeverityString(defaultSeverity)
                            	: GetSeverityString(null);
```

Comment (Evangelink) [reply src/CodeStyle/Tools/Program.cs:198]: or

```suggestion
                            return GetSeverityString(isEnabledByDefault ? defaultSeverity : null);
```

File: src/CodeStyle/Tools/Program.cs [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+                                    return "none";
+                                }
+
+                                return severity.Value switch
+                                {
+                                    DiagnosticSeverity.Error => "error",
+                                    DiagnosticSeverity.Warning => "warning",
+                                    DiagnosticSeverity.Info => "suggestion",
+                                    DiagnosticSeverity.Hidden => "silent",
+                                    _ => throw new NotImplementedException(severity.Value.ToString()),
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:213]: Maybe `NotSupportedException` or `ArgumentOutOfRangeException`.

File: src/CodeStyle/Tools/Program.cs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+            private AnalyzerAssemblyLoader() { }
+            public void AddDependencyLocation(string fullPath)
+            {
+            }
+
+            public Assembly LoadFromPath(string fullPath)
+            {
+                return Assembly.LoadFrom(fullPath);
+            }
```
Comment (Evangelink) [inline src/CodeStyle/Tools/Program.cs:299]:
```suggestion
	            => Assembly.LoadFrom(fullPath);
```

File: src/CodeStyle/Tools/Program.cs [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+            Minimum,
+            Recommended,
+            All
+        }
+
+        private sealed class AnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
+        {
+            public static IAnalyzerAssemblyLoader Instance = new AnalyzerAssemblyLoader();
+
+            private AnalyzerAssemblyLoader() { }
```
Comment (Youssef1313) [inline src/CodeStyle/Tools/Program.cs:291]:
```suggestion
            private AnalyzerAssemblyLoader() { }

```

File: src/CodeStyle/Tools/Program.cs (hunk 11 above)
Comment (Youssef1313) [inline src/CodeStyle/Tools/Program.cs:213]: This is a duplication from an existing internal extension method. I'm not sure if it's okay to link to a file in the workspaces layer from here?

**Review by Youssef1313 — COMMENTED** [review]
nits

File: src/CodeStyle/Tools/Program.cs (hunk 1 above)
Comment (mavasani) [reply src/CodeStyle/Tools/Program.cs:38]: I have cloned this code from analyzers repo, so tried to keep it as much consistent as possible. I can fix this suggestion in both repos though.

File: src/CodeStyle/Tools/Program.cs (hunk 3 above)
Comment (mavasani) [reply src/CodeStyle/Tools/Program.cs:94]: I personally like local functions of local functions and feel it avoids polluting private code used by local functions. Hence, they were added to roslyn-analyzers repo and I retained most of the code structure as-is during the port. I can change that if others disagree. I'll wait to hear from couple more folks.

File: src/Analyzers/Core/Analyzers/EnforceOnBuild.cs [hunk 14]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    /// <summary>
+    /// Build enforcement recommendation for a code style analyzer.
+    /// </summary>
+    internal enum EnforceOnBuild
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Analyzers/Core/Analyzers/EnforceOnBuild.cs:10]: Renaming the enum or fields should be fine as the names themselves are not emitted into any of the generated files. The same file is linked into both CodeStyle project and the post-build tool, so the tool will also be correctly updated when renames happen.

File: src/CodeStyle/Tools/Program.cs [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace CodeStyleConfigFileGenerator
+{
+    public static class Program
+    {
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/CodeStyle/Tools/Program.cs:20]: I did that first, but this type has nested enum and a class, which needs it to be inside another class. I did not want to make them internal.

File: src/CodeStyle/Tools/Program.cs [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+
+            CreateGlobalConfigFiles(language, outputDir, assemblyList);
+            CreateTargetsFile(language, outputDir, targetsFileName);
+            return 0;
+        }
+
+        private static void CreateGlobalConfigFiles(string language, string outputDir, List<string> assemblyList)
+        {
+            Debug.Assert(language is "CSharp" or "VisualBasic");
+            var languageForGetAnalyzers = language == "CSharp" ? LanguageNames.CSharp : LanguageNames.VisualBasic;
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/CodeStyle/Tools/Program.cs:48]: Yes, there are differences between input values asserted in the above line and LanguageName constants: https://github.com/dotnet/roslyn/blob/2c3fb8fb367b4499731b89fdd041c512d262f020/src/Compilers/Core/Portable/Symbols/LanguageNames.cs#L15-L20

**Review by Evangelink — APPROVED** [review]

**Comment by mavasani** [comment]
Thanks @Evangelink!

@dotnet/roslyn-ide for another review.

File: Compilers.sln [hunk 17]
```diff
@@ -127,10 +127,16 @@ EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.CSharp.CodeStyle", "src\CodeStyle\CSharp\Analyzers\Microsoft.CodeAnalysis.CSharp.CodeStyle.csproj", "{AA87BFED-089A-4096-B8D5-690BDC7D5B24}"
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes", "src\CodeStyle\CSharp\CodeFixes\Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes.csproj", "{A07ABCF5-BC43-4EE9-8FD8-B2D77FD54D73}"
+	ProjectSection(ProjectDependencies) = postProject
```
Comment (Youssef1313) [inline Compilers.sln:130]: Are there any changes needed in the solution filter (`Compilers.slnf`) file?

There is an on-going work by @sharwell (https://github.com/dotnet/roslyn/pull/47531) to remove Compilers.sln.

**Review by jmarolf — APPROVED** [review]
:shipit:

**Pull request merged**